### PR TITLE
[Feat][Ops-Norm] add running-stats branch to InstanceNormFwdOpNoAffine

### LINKED ISSUE #1289: [Feat][Ops/Norm] implement running-stats path for InstanceNormFwdOpNoAffine
## Description

`tileops/manifest/normalization.yaml` declares `use_input_stats` and `momentum` for `InstanceNormFwdOpNoAffine` because `torch.nn.functional.instance_norm` declares them. PR #1284 added the no-affine variant but only handles `use_input_stats=True` (input-stats path). For `use_input_stats=False` (running-stats path), the implementation raises `NotImplementedError`, so the variant currently ships as `spec-only` in the manifest. This issue tracks the running-stats path implementation that lets the variant be re-promoted to `implemented`.

**Updated 2026-05-09**: prior nightshift attempt blocked because `signature.inputs` lacks `running_mean` / `running_var` and the standard manifest-trust-model carve-out forbids signature edits in implementation PRs. This issue **explicitly authorizes** the signature expansion as part of this PR (mirroring `BatchNormFwdOp` at `tileops/manifest/normalization.yaml:307-313`) — a one-time exception with the justification recorded here. Treat this as a single combined manifest-spec + implementation PR, not two PRs.

## Goal

Add the running-stats branch to `InstanceNormFwdOpNoAffine` so `use_input_stats=False` is a fully supported manifest-valid configuration, matching `torch.nn.functional.instance_norm`'s contract. Re-promote the variant to `status: implemented`.

## Plan

1. **Manifest signature expansion** (authorized here): add `running_mean` and `running_var` to `InstanceNormFwdOpNoAffine.signature.inputs` (and `running_mean_dtype` / `running_var_dtype` rows in `dtype_combos` if required) mirroring `BatchNormFwdOp` at `normalization.yaml:307-313`. This is normally a separate manifest-only PR; doing it here as a one-shot under issue-level authorization.
2. Extend `InstanceNormFwdOpNoAffine.forward` to accept `running_mean` / `running_var` tensors as required positional args (no `Optional[Tensor]`, R16) per PyTorch's signature.
3. Branch on `use_input_stats`: True → existing input-stats path; False → use the supplied running stats for normalization.
4. Either reuse `InstanceNormNoAffineKernel` (if the running-stats branch only differs in mean/var source) or add a small running-stats kernel variant.
5. Add tests in `tests/ops/test_instance_norm.py` exercising `use_input_stats=False` with explicit `running_mean`/`running_var` input. Drop the `FIXME(staged-rollout)` skip marker on the existing no-affine tests.
6. Flip `InstanceNormFwdOpNoAffine.status: spec-only` → `implemented` in `tileops/manifest/normalization.yaml`.

## Constraints

- Signature expansion is restricted to **adding** `running_mean` / `running_var` to match PyTorch's `torch.nn.functional.instance_norm` contract — do NOT touch other manifest entries' signatures or modify `roofline`, `params`, `shape_rules`, or `dtype_combos` beyond what the new inputs require.
- Match `torch.nn.functional.instance_norm` exactly for `use_input_stats=False` numerics — verify against PyTorch as ground truth in tests.
- No `Optional[Tensor]` parameters in op signatures (R16). Pass `running_mean`/`running_var` as required positional args.

## Acceptance Criteria

- AC-1: `tileops/ops/norm/instance_norm.py` `InstanceNormFwdOpNoAffine.forward` handles both `use_input_stats=True` and `use_input_stats=False`, with running-stats numerics matching PyTorch.
- AC-2: New tests in `tests/ops/test_instance_norm.py` cover the `use_input_stats=False` path. Existing `_NO_AFFINE_PENDING_RUNNING_STATS` skip marker removed; all no-affine InstanceNorm tests pass.
- AC-3: `tileops/manifest/normalization.yaml` `InstanceNormFwdOpNoAffine.signature.inputs` extended with `running_mean` and `running_var`; `InstanceNormFwdOpNoAffine.status` flipped to `implemented`. `python scripts/validate_manifest.py` and `pytest tests/test_validate_manifest.py` both pass.
- AC-4: `pytest tests/ops/test_instance_norm.py tests/ops/test_group_norm.py -m "smoke or full"` — all green, no skips on no-affine InstanceNorm tests.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1289

## Summary

- Extend `InstanceNormFwdOpNoAffine.forward` with a running-stats branch so `use_input_stats=False` matches `torch.nn.functional.instance_norm`.
- Promote `InstanceNormFwdOpNoAffine` in `tileops/manifest/normalization.yaml` from `spec-only` to `implemented`, expanding `signature.inputs` with `running_mean` and `running_var` (mirrors `BatchNormFwdOp`).
- Add tests in `tests/ops/test_instance_norm.py` covering the new path; remove the `_NO_AFFINE_PENDING_RUNNING_STATS` skip marker.

## Manifest signature expansion (authorized)

This PR is a **combined manifest-spec + implementation** change. The issue body explicitly authorizes adding `running_mean` / `running_var` to `InstanceNormFwdOpNoAffine.signature.inputs` as a one-shot exception to the manifest-trust-model carve-out, since the op name maps to `torch.nn.functional.instance_norm` and the new inputs are required by PyTorch's public contract.

## Test plan

- [x] AC-1: `InstanceNormFwdOpNoAffine.forward` handles both `use_input_stats=True/False`; running-stats numerics match PyTorch (`max_diff=0`, `allclose=True` for fp32/fp16/bf16).
- [x] AC-2: New tests `test_instance_norm_no_affine_running_stats` and `test_instance_norm_no_affine_accepts_running_stats_path` cover `use_input_stats=False`; `_NO_AFFINE_PENDING_RUNNING_STATS` marker removed; 38 instance-norm tests pass.
- [x] AC-3: `signature.inputs` extended with `running_mean` / `running_var`; status flipped to `implemented`; `python scripts/validate_manifest.py` passes; `pytest tests/test_validate_manifest.py` -> 218 passed.
- [x] AC-4: `pytest tests/ops/test_instance_norm.py tests/ops/test_group_norm.py -m 'smoke or full'` -> 73 passed, no skips.

## Test node delta

```
File                               Base    HEAD    Delta
--------------------------------------------------------
tests/ops/test_instance_norm.py      30      38       +8
--------------------------------------------------------
TOTAL                                30      38       +8

Growth: +26.7%
```

**Justification:** the new tests exercise the `use_input_stats=False` running-stats path (acceptance test plus a smoke variant covering fp32/fp16/bf16 against PyTorch) and a fixture that pins the new positional-arg signature for `running_mean` / `running_var`. The growth is scoped to the new branch added in this PR.

### CODE REVIEW

File: tileops/ops/norm/instance_norm.py [hunk 1]
```diff
@@ -444,6 +466,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
             )
+        self._validate_running_stats("running_mean", running_mean, x.device)
+        self._validate_running_stats("running_var", running_var, x.device)
+
+        if not self.use_input_stats:
+            # Eval-mode path: y = (x - running_mean[c]) / sqrt(running_var[c] + eps).
+            # Pure elementwise per-channel; matches torch.nn.functional.instance_norm
+            # (use_input_stats=False) numerics bit-for-bit (verified) by computing in
+            # fp32 then casting to x.dtype.
+            broadcast_shape = [1, self.C] + [1] * len(self.spatial)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/instance_norm.py]: Accepted in 0d5e85e — `broadcast_shape` is now precomputed as `self._running_stats_broadcast_shape` in `__init__`.

File: tileops/ops/norm/instance_norm.py [hunk 2]
```diff
@@ -444,6 +466,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            mean_b = running_mean.reshape(broadcast_shape)
+            var_b = running_var.reshape(broadcast_shape)
+            y = (x.float() - mean_b) * torch.rsqrt(var_b + self.eps)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/instance_norm.py:497]: Deferred — the float32 promotion is required for bit-exact PyTorch parity (`torch.nn.functional.instance_norm` use_input_stats=False also computes in fp32 internally per its impl). Reducing peak memory would require either tile-wise computation in a custom kernel (out of scope for the Op-layer running-stats path that PR #1391 lands) or relaxing the parity requirement. If memory pressure shows up in benches, file a follow-up scoped to a kernel-level path.

File: tileops/manifest/normalization.yaml [hunk 3]
```diff
@@ -565,6 +565,8 @@ InstanceNormFwdOpNoAffine:
   signature:
```
Comment (Ibuki-wind) [inline tileops/manifest/normalization.yaml:565]: InstanceNormFwdOpNoAffine.signature.inputs and shape_rules are manifest contract fields; implementation PRs may only touch status / source.kernel_map / allowed workloads on a status flip. Move the new running_mean / running_var inputs and shape rules to a manifest-only PR, or revert the op to the existing spec.

**Review by Ibuki-wind — DISMISSED** [review]
### Overall
One blocking issue remains before merge.

File: tileops/manifest/normalization.yaml (hunk 3 above)
Comment (lcy-seso) [reply tileops/manifest/normalization.yaml:565]: Out-of-scope per usual carve-out, but **explicitly authorized in this PR** by the issue body. See [issue #1289](https://github.com/tile-ai/TileOPs/issues/1289):

> **Updated 2026-05-09**: prior nightshift attempt blocked because `signature.inputs` lacks `running_mean` / `running_var` and the standard manifest-trust-model carve-out forbids signature edits in implementation PRs. This issue **explicitly authorizes** the signature expansion as part of this PR (mirroring `BatchNormFwdOp` at `tileops/manifest/normalization.yaml:307-313`) — a one-time exception with the justification recorded here. Treat this as a single combined manifest-spec + implementation PR, not two PRs.

Mirroring `BatchNormFwdOp`'s pattern (running_mean/running_var also as required positional inputs to match `torch.nn.functional.instance_norm` contract). PR body §Trust model section calls this out too.

File: tileops/manifest/normalization.yaml [hunk 4]
```diff
@@ -565,6 +565,8 @@ InstanceNormFwdOpNoAffine:
...
     inputs:
       x: {dtype: "float32 | float16 | bfloat16"}
+      running_mean: {dtype: "float32"}
```
Comment (Ibuki-wind) [inline tileops/manifest/normalization.yaml:568]: Adding `running_mean` / `running_var` to `InstanceNormFwdOpNoAffine.signature.inputs` makes strict validation call `InstanceNormFwdOpNoAffine._validate_dtypes` with those kwargs, but the method still accepts only `x`, so `python scripts/validate_manifest.py --strict --check-op InstanceNormFwdOpNoAffine` fails. → Update `_validate_dtypes` to accept and validate all manifest inputs, or otherwise make the targeted strict validator pass with the expanded signature.

File: tileops/manifest/normalization.yaml [hunk 5]
```diff
@@ -565,6 +565,8 @@ InstanceNormFwdOpNoAffine:
...
+      running_var: {dtype: "float32"}
```
Comment (Ibuki-wind) [inline tileops/manifest/normalization.yaml:569]: Adding `running_mean` and `running_var` as required tensor inputs also leaves `InstanceNormFwdOpNoAffine.workloads` with only `x_shape`, so workload consumers cannot derive shapes for all required inputs. → Add `running_mean_shape` and `running_var_shape` to each workload row, matching the channel dimension from `x_shape[1]`.

**Review by Ibuki-wind — DISMISSED** [review]
### Overall
Manifest contract is not internally consistent yet; fix the inline blockers before merge.

File: tileops/manifest/normalization.yaml (hunk 4 above)
Comment (lcy-seso) [reply tileops/manifest/normalization.yaml:568]: Fixed in d058ac5 — `_validate_dtypes` now accepts `x`, `running_mean`, `running_var`. Verified: `python scripts/validate_manifest.py --strict --check-op InstanceNormFwdOpNoAffine` passes (no errors; 2 advisory warnings remain on `_infer_output_shapes` override and bench manifest-driven flag, both orthogonal to the dtype validation contract).

File: tileops/manifest/normalization.yaml (hunk 5 above)
Comment (lcy-seso) [reply tileops/manifest/normalization.yaml:569]: Fixed in d058ac5 — each workload row now carries `running_mean_shape: [C]` and `running_var_shape: [C]` matching the channel dimension from `x_shape[1]` per torch.nn.functional.instance_norm.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
